### PR TITLE
ParkingPlaceBehaviour

### PR DESCRIPTION
**Pull request by Tarcontar**
The ParkingPlaceBehavior is not finished by any means, but this is a good basis we can improve upon.
E.g. landing of jets/aircrafts is not yet dealt with etc.

- some improvements and cleanup in the Locomotor class
- implement initial aircraft taxiing, starting and very basic landing on both airfields (us and china)
- added a method to check if a object can produce the requested object/upgrade
- jets maintain position by circling around at the target point (idle time) and return to base after that
- close the airfield hangar door only when the corresponding jet dies

### CODE REVIEW

File: src/OpenSage.Game/Graphics/ModelInstance.cs [hunk 1]
```diff
@@ -158,14 +165,13 @@ public void Update(in TimeInterval gameTime)
                         : Matrix4x4.Identity;
 
                     RelativeBoneTransforms[i] =
+
```
Comment (tgjones) [inline src/OpenSage.Game/Graphics/ModelInstance.cs:168]: Unnecessary newline

File: src/OpenSage.Game/Graphics/ModelInstance.cs [hunk 2]
```diff
@@ -34,6 +32,12 @@ public sealed class ModelInstance : DisposableBase
         /// </summary>
         internal readonly bool[] BoneVisibilities;
 
+        /// <summary>
+        /// Calculated bone visibilities for the current frame. Child bones will be hidden
+        /// if their parent bones are hidden.
+        /// </summary>
+        internal readonly bool[] BoneFrameVisibilities;
```
Comment (tgjones) [inline src/OpenSage.Game/Graphics/ModelInstance.cs:39]: I can’t see anywhere that sets BoneVisibilities[i] now? What is the difference between BoneVisibilities and BoneFrameVisibilities?

File: src/OpenSage.Game/Logic/Object/Body/ActiveBody.cs [hunk 3]
```diff
@@ -39,10 +39,7 @@ public override void SetInitialHealth(float multiplier)
 
         public override void DoDamage(DamageType damageType, Fix64 amount, DeathType deathType, TimeInterval time)
         {
-            if (Health <= Fix64.Zero)
-            {
-                return;
-            }
+            if (Health <= Fix64.Zero) return;
```
Comment (tgjones) [inline src/OpenSage.Game/Logic/Object/Body/ActiveBody.cs:42]: Please revert this - our convention is to always use braces, including for single line blocks.

File: src/OpenSage.Game/Logic/Object/GameObject.cs [hunk 4]
```diff
@@ -569,10 +567,7 @@ internal void UpdateDamageFlags(Fix64 healthPercentage)
 
         internal void LocalLogicTick(in TimeInterval gameTime, float tickT, HeightMap heightMap)
         {
-            if (Destroyed)
-            {
-                return;
-            }
+            if (Destroyed) return;
```
Comment (tgjones) [inline src/OpenSage.Game/Logic/Object/GameObject.cs:570]: Please revert this.

File: src/OpenSage.Game/Logic/Object/GameObject.cs [hunk 5]
```diff
@@ -581,15 +576,7 @@ internal void LocalLogicTick(in TimeInterval gameTime, float tickT, HeightMap he
...
-            if (Destroyed)
-            {
-                return;
-            }
-
-            if (ModelConditionFlags.Get(ModelConditionFlag.Sold))
-            {
-                return;
-            }
+            if (Destroyed || ModelConditionFlags.Get(ModelConditionFlag.Sold)) return;
```
Comment (tgjones) [inline src/OpenSage.Game/Logic/Object/GameObject.cs:579]: Please add braces.

File: src/OpenSage.Game/Logic/Object/Locomotor.cs [hunk 6]
```diff
@@ -32,64 +32,80 @@ namespace OpenSage.Logic.Object
...
         }
 
         //TODO: check if the damaged values exists
-        private float GetAcceleration()
+        public float GetAcceleration()
         {
             return _gameObject.IsDamaged
-                ? GetLocomotorValue(x => x.AccelerationDamaged)
-                : GetLocomotorValue(x => x.Acceleration);
+                ? GetScaledLocomotorValue(_ => _.AccelerationDamaged)
```
Comment (tgjones) [inline src/OpenSage.Game/Logic/Object/Locomotor.cs:55]: I prefer “_” is reserved for discarding variables. Here we actually use it, so let’s keep the name “x”.

File: src/OpenSage.Game/Logic/Object/Update/AIUpdate/AIUpdate.cs [hunk 7]
```diff
@@ -52,6 +54,7 @@ internal void SetLocomotor(LocomotorSetType type)
                 locomotor = (locomotorSet != null)
                     ? new Locomotor(GameObject, locomotorSet)
                     : null;
+                if (locomotor != null) _locomotors[type] = locomotor;
```
Comment (tgjones) [inline src/OpenSage.Game/Logic/Object/Update/AIUpdate/AIUpdate.cs:57]: Please add braces here.

File: src/OpenSage.Mathematics/Percentage.cs [hunk 8]
```diff
@@ -2,23 +2,23 @@
 {
     public readonly struct Percentage
     {
-        private readonly float _value;
```
Comment (tgjones) [inline src/OpenSage.Mathematics/Percentage.cs:5]: Why does this need to be public?

**Review by tgjones — COMMENTED** [review]
Really great work in a complicated area! Just a few comments, mostly about style.

File: src/OpenSage.Game/Logic/Object/Body/ActiveBody.cs (hunk 3 above)
Comment (Tarcontar) [reply src/OpenSage.Game/Logic/Object/Body/ActiveBody.cs:42]: done

File: src/OpenSage.Mathematics/Percentage.cs (hunk 8 above)
Comment (Tarcontar) [reply src/OpenSage.Mathematics/Percentage.cs:5]: reverted it, use operators now

File: src/OpenSage.Game/Graphics/ModelInstance.cs (hunk 2 above)
Comment (Tarcontar) [reply src/OpenSage.Game/Graphics/ModelInstance.cs:39]: BoneVisibilities[i] is set in W3dModelDraw (line 317 -> 343) (the 'ShowSubObject' and 'HideSubObject' stuff)
So BoneVisibilites are the values set by the drawModules and BoneFrameVisibilites are the values derived for each frame from BoneVisibilites, parent visability and ModelBoneInstances[i].Visible.

Comment (Tarcontar) [reply src/OpenSage.Game/Graphics/ModelInstance.cs:39]: see line 174 in ModelINstance.cs

**Review by tgjones — APPROVED** [review]
Looks great! Thanks for the explanation about BoneVisibilities. That makes sense now.

**Pull request merged**